File: frontend/src/types/exam.ts
export interface Topic {
  id: string
  name: string
}

export interface Subject {
  id: string
  name: string
  order_index: number
  topics: Topic[]
}

export interface Exam {
  id: string
  title: string
  description: string | null
  exam_type: string
  is_published: boolean
  created_by: string
  created_at: string
  subjects?: Subject[] | null
}

export interface ExamCreatePayload {
  title: string
  exam_type: string
  description?: string
}

export interface ExamUpdatePayload {
  title?: string
  description?: string
  is_published?: boolean
}

export interface SubjectCreatePayload {
  name: string
  order_index?: number
}

export interface Paginated<T> {
  total: number
  page: number
  limit: number
  items: T[]
}

File: backend/app/schemas/exam.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ─── Topic ──────────────────────────────────────────────
class TopicCreate(BaseModel):
    name: str


class TopicOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str


# ─── Subject ────────────────────────────────────────────
class SubjectCreate(BaseModel):
    name: str
    order_index: int = 0


class SubjectOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    order_index: int
    topics: list[TopicOut] = []


# ─── Exam ───────────────────────────────────────────────
class ExamCreate(BaseModel):
    title: str
    exam_type: str
    description: str | None = None


class ExamUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    is_published: bool | None = None


class ExamOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    description: str | None = None
    exam_type: str
    is_published: bool
    created_by: UUID
    created_at: datetime
    subjects: list[SubjectOut] | None = None
